### PR TITLE
feat(VDatePicker): pick date range

### PR DESCRIPTION
**Pull request by YipingRuan**
#1646

## Description

Allow datepicker to pick range.
By add new prop "Range", a workaround is provided to handle range picking.

- Value will be kept as an array of size 2 to save [From, To].  New click will clear the array
- Any dates between will be displayed as "selected"
- Can select with both direction (From then To, or To then From)
## Motivation and Context

[#1646 ](https://github.com/vuetifyjs/vuetify/issues/1646)
## How Has This Been Tested?

visually and unit
VDatePicker

## Markup:

<details>

```vue
<template>
  <v-container>
    <v-date-picker v-model="dates" range width="290" class="mt-4"></v-date-picker>
    v-model value: {{ dates }}
    <v-menu ref="menu" v-model="menu" :close-on-content-click="false" transition="scale-transition" offset-y min-width="290px">
      <template v-slot:activator="{ on }">
        <v-text-field :value="dateRangeText" label="Date range" prepend-icon="event" readonly v-on="on"></v-text-field>
      </template>
      <v-date-picker v-model="dates" range no-title scrollable>
        <div class="flex-grow-1"></div>
        <v-btn text color="primary" @click="$refs.menu.save(dates)">OK</v-btn>
      </v-date-picker>
    </v-menu>
  </v-container>
</template>

<script>
  export default {
    data: () => ({
      dates: ['2018-09-15', '2018-09-20'],
      menu: false,
    }),
    computed: {
      dateRangeText () {
        return this.dates.join(' ~ ')
      },
    },
  }
</script>

```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [X] The PR title is no longer than 64 characters.
- [X] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and breaking changes).
- [X] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)
- [x] I've added new examples to the kitchen (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/components/VDatePicker/VDatePicker.ts [hunk 1]
```diff
@@ -90,6 +90,8 @@ export default mixins(
       type: String,
       default: '$vuetify.datePicker.itemsSelected',
     },
+    // Allow select date with range, must be used together with multiple
+    selectRange: Boolean,
```
Comment (sh7dm) [inline packages/vuetify/src/components/VDatePicker/VDatePicker.ts]:
```suggestion
    range: Boolean,
```

Comment (sh7dm) [inline packages/vuetify/src/components/VDatePicker/VDatePicker.ts]: We can enable multiple automatically if range

Comment (sh7dm) [reply packages/vuetify/src/components/VDatePicker/VDatePicker.ts]: Anyway, comment isn't very useful here
```suggestion
```

File: packages/vuetify/src/components/VDatePicker/__tests__/VDatePicker.date.spec.ts [hunk 2]
```diff
@@ -633,4 +633,26 @@ describe('VDatePicker.ts', () => { // eslint-disable-line max-statements
...
+    })
+
+    wrapper.vm.$on('input', cb)
+    wrapper.findAll('.v-date-picker-table--date tbody tr+tr td:first-child button').at(0).trigger('click')
+    expect(cb.mock.calls[0][0]).toEqual(
+      expect.arrayContaining(['2019-01-06'])
+    )
+
+    wrapper.findAll('.v-date-picker-table--date tbody tr+tr td button').at(2).trigger('click')
+    console.log(JSON.stringify(cb.mock.calls))
```
Comment (sh7dm) [inline packages/vuetify/src/components/VDatePicker/__tests__/VDatePicker.date.spec.ts]: Remove logs from any non-WIP tests
```suggestion
```

File: packages/vuetify/src/components/VDatePicker/__tests__/VDatePicker.date.spec.ts [hunk 3]
```diff
@@ -633,4 +633,26 @@ describe('VDatePicker.ts', () => { // eslint-disable-line max-statements
...
+    expect(cb.mock.calls[0][0][0]).toBe('2019-01-06')
+    expect(cb.mock.calls[1][0][0]).toBe('2019-01-08')
```
Comment (sh7dm) [inline packages/vuetify/src/components/VDatePicker/__tests__/VDatePicker.date.spec.ts]: why? first date should equal `2019-01-01`, shouldn't it?

Comment (YipingRuan) [reply packages/vuetify/src/components/VDatePicker/__tests__/VDatePicker.date.spec.ts]: I am trying to use the array of length 2 to keep the [From ,To] range. So further clicking will wipe the previous range.
['2019-01-01', '2019-01-02'] => first click => ['2019-01-06'] => second click = ['2019-01-06', '2019-01-08']

File: packages/vuetify/src/components/VDatePicker/VDatePicker.ts (hunk 1 above)
Comment (YipingRuan) [reply packages/vuetify/src/components/VDatePicker/VDatePicker.ts]: Changed to Range and use computed internalMultiple

File: packages/vuetify/src/components/VDatePicker/VDatePicker.ts [hunk 4]
```diff
@@ -122,21 +121,24 @@ export default mixins(
...
-        const date = (this.multiple ? (this.value as string[])[(this.value as string[]).length - 1] : this.value) ||
+        const date = (this.multiple || this.range ? (this.value as string[])[(this.value as string[]).length - 1] : this.value) ||
           `${now.getFullYear()}-${now.getMonth() + 1}`
         return sanitizeDateString(date as string, this.type === 'date' ? 'month' : 'year')
       })(),
     }
   },
 
   computed: {
+    internalMultiple (): boolean {
```
Comment (sh7dm) [inline packages/vuetify/src/components/VDatePicker/VDatePicker.ts]: probably `isMultiple` will be better

Comment (YipingRuan) [reply packages/vuetify/src/components/VDatePicker/VDatePicker.ts]: Should I put a comment there to explain this compute, like:
// Range selection automatically indicate multiple

Comment (sh7dm) [reply packages/vuetify/src/components/VDatePicker/VDatePicker.ts]: no, just rename.

File: packages/docs/src/examples/date-pickers/intermediate/date-range.vue [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  </v-row>
+</template>
+
+<script>
+  export default {
+    data: () => ({
+      dates: ['2019-09-10', '2019-09-20'],
+    }),
+    computed: {
+      dateRangeText: {
```
Comment (sh7dm) [inline packages/docs/src/examples/date-pickers/intermediate/date-range.vue]: just use getter as a computed itself:
```ts
computed: {
  dateRangeText () {
    return this.dates.join(' ~ ')
  }
}
```

File: packages/docs/src/lang/en/components/DatePickers.json [hunk 6]
```diff
@@ -44,6 +44,10 @@
       "desc": "Date picker can now select multiple dates with the `multiple` prop. If using `multiple` then date picker expects its model to be an array.",
       "header": "### Date pickers - Mutiple"
     },
+    "date-range": {
+      "desc": "Date picker can select date range with the `range` property. When using `range` property date picker expects its model to be an array of length 2 or empty.",
```
Comment (sh7dm) [inline packages/docs/src/lang/en/components/DatePickers.json]: Use `prop` for consistency here please.

```suggestion
      "desc": "Date picker can select date range with the `range` prop. When using `range` prop date picker expects its model to be an array of length 2 or empty.",
```

File: packages/docs/src/examples/date-pickers/intermediate/date-range.vue (hunk 5 above)
Comment (YipingRuan) [reply packages/docs/src/examples/date-pickers/intermediate/date-range.vue]: Updated in new commit

**Pull request merged**